<a href="https://colab.research.google.com/github/eduardopadilla3/DS-Unit-1-Sprint-2-Statistics/blob/master/module2/Copy_of_LS_DS_122_Chi2_Tests_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Assignment - Practice Chi^2 Tests

Use the following dataset relating to math scores of students in two different Portugese schools:

<https://archive.ics.uci.edu/ml/datasets/Student+Performance>

### 1) Load the dataset specifically relating to math scores as a new dataframe.

There are two datasets in the `student.zip` file, make sure you use `student-mat.csv`.


In [0]:
import pandas as pd
import numpy as np
from scipy import stats

In [0]:
# YOUR WORK HERE
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip

--2020-06-09 20:34:58--  https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 20478 (20K) [application/x-httpd-php]
Saving to: ‘student.zip’

student.zip         100%[===================>]  20.00K  --.-KB/s    in 0.1s    

2020-06-09 20:34:59 (154 KB/s) - ‘student.zip’ saved [20478/20478]



In [0]:
!unzip student.zip

Archive:  student.zip
  inflating: student-mat.csv         
  inflating: student-por.csv         
  inflating: student-merge.R         
  inflating: student.txt             


In [0]:
df = pd.read_csv('student-mat.csv', sep = ';')

df.head()

school sex  age address famsize Pstatus  ...  Walc  health absences  G1  G2  G3
0     GP   F   18       U     GT3       A  ...     1       3        6   5   6   6
1     GP   F   17       U     GT3       T  ...     1       3        4   5   5   6
2     GP   F   15       U     LE3       T  ...     3       3       10   7   8  10
3     GP   F   15       U     GT3       T  ...     1       5        2  15  14  15
4     GP   F   16       U     GT3       T  ...     2       5        4   6  10  10

[5 rows x 33 columns]

### 2) Use Chi^2 tests and `stats.chi2_contingency()` to identify:
 - Two pairs of variables that are dependent (are associated with one another).
 - Two pairs of variables that are independent (have no significant relationship).

Does it make intuitive sense why the variables in these pairs might or might not show a relationship to one another? 


In [0]:
# YOUR WORK HERE
df['Pstatus'].value_counts().sort_index()

A     41
T    354
Name: Pstatus, dtype: int64

In [0]:
df['romantic'].value_counts().sort_index()

no     263
yes    132
Name: romantic, dtype: int64

In [0]:
contingency = pd.crosstab(df['Pstatus'], df['romantic'])

contingency

romantic   no  yes
Pstatus           
A          25   16
T         238  116

In [0]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingency)

print("chi2 statistic", chi2)
print("p value", p_value)
print("degrees of freedom",dof)
print("expected frequencies table", expected)

chi2 statistic 0.39573864435183936
p value 0.5292982453234916
degrees of freedom 1
expected frequencies table [[ 27.29873418  13.70126582]
 [235.70126582 118.29873418]]


Null Hypothesis: Pstatus and romantic are independent 

Alternative Hypothesis: Pstatus and romantic are dependent (associated)

Confidence Level: 95%

Conclusion:
Based on a chi2 statistic of 0.39, and a p-value of 0.529, we fail to reject the null hypothesis that Pstatus and romantic are independent. There is no relationship

In [0]:
df['famsize'].value_counts().sort_index()

GT3    281
LE3    114
Name: famsize, dtype: int64

In [0]:
df['Mjob'].value_counts().sort_index()

at_home      59
health       34
other       141
services    103
teacher      58
Name: Mjob, dtype: int64

In [0]:
contingency = pd.crosstab(df['famsize'], df['Mjob'])

contingency

Mjob     at_home  health  other  services  teacher
famsize                                           
GT3           44      24    106        70       37
LE3           15      10     35        33       21

In [0]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingency)

print("chi2 statistic", chi2)
print("p value", p_value)
print("degrees of freedom",dof)
print("expected frequencies table", expected)

chi2 statistic 3.4955222729693007
p value 0.47855951929693274
degrees of freedom 4
expected frequencies table [[ 41.9721519   24.18734177 100.30632911  73.27341772  41.26075949]
 [ 17.0278481    9.81265823  40.69367089  29.72658228  16.73924051]]


Null Hypothesis: famsize and mjob are independent

Alternative Hypothesis: famsize and mjob are dependent (associated)

Confidence Level: 95%

Conclusion:
Based on chi2 value of 3.495 and pvalue of 0.4785, we reject the null hypothesis and thus conclude that family size and mother's job have some kind of relationship

### 3) Use NumPy to perform your own chi^2 test "from scratch" 

Pick any of the chi^2 tests that you ran in part 2 and try them on your own. You should get the same results that Scipy got for all four values returned from `chi2_contingency()`

In [0]:
# YOUR WORK HERE
#observed:
observed = pd.crosstab(df['famsize'], df['Mjob'])

observed

Mjob     at_home  health  other  services  teacher
famsize                                           
GT3           44      24    106        70       37
LE3           15      10     35        33       21

In [0]:
#expected:
observed_with_margins = pd.crosstab(df['famsize'], df['Mjob'], margins=True)

observed_with_margins

Mjob     at_home  health  other  services  teacher  All
famsize                                                
GT3           44      24    106        70       37  281
LE3           15      10     35        33       21  114
All           59      34    141       103       58  395

In [0]:
row_totals = df['famsize'].value_counts().sort_index().values

row_totals

array([281, 114])

In [0]:
col_totals = df['Mjob'].value_counts().sort_index().values

col_totals

array([ 59,  34, 141, 103,  58])

In [0]:
sample_size = df.shape[0]

sample_size

395

In [0]:
my_list = []

for i, row_total in enumerate(row_totals):
  row = []
  for j, column_total in enumerate(col_totals):
    expected_val  = ((column_total*row_total) / sample_size)
    row.append(expected_val)
  my_list.append(row)

my_list

[[41.972151898734175,
  24.1873417721519,
  100.30632911392405,
  73.27341772151898,
  41.26075949367089],
 [17.02784810126582,
  9.812658227848102,
  40.69367088607595,
  29.72658227848101,
  16.739240506329114]]

In [0]:
expected = my_list

expected

[[41.972151898734175,
  24.1873417721519,
  100.30632911392405,
  73.27341772151898,
  41.26075949367089],
 [17.02784810126582,
  9.812658227848102,
  40.69367088607595,
  29.72658227848101,
  16.739240506329114]]

In [0]:
observed = observed.values
observed

array([[ 44,  24, 106,  70,  37],
       [ 15,  10,  35,  33,  21]])

In [0]:
((observed - expected)**2 / expected).sum()

3.4955222729693007

In [0]:
#dof:
(2-1)*(5-1)

4

## Stretch goals:

### 1. Refactor your code so it is elegant, readable, and holds reusable code in functions.

In [0]:
# YOUR WORK HERE



### 2. Check For Understanding - Study and write your own explanations/definitions for these topics:

- What is a sample "estimate" in statistics?

- What are hypothesis test? How are they useful?

- What is a "null hypothesis?"

- What is a p-value and what does it represent?

- What does it mean for something to be "statistically significant?"

- What is a test statistic and how does it relate to a p-value?

- What are "degrees of freedom" and how are they calculated in a 1-sample, 2-sample, and chi^2 test for independence?

## Resources

- [Interactive visualize the Chi-Squared test](https://homepage.divms.uiowa.edu/~mbognar/applets/chisq.html)
- [Calculation of Chi-Squared test statistic](https://en.wikipedia.org/wiki/Pearson%27s_chi-squared_test)
- [Visualization of a confidence interval generated by R code](https://commons.wikimedia.org/wiki/File:Confidence-interval.svg)
- [Expected value of a squared standard normal](https://math.stackexchange.com/questions/264061/expected-value-calculation-for-squared-normal-distribution) (it's 1 - which is why the expected value of a Chi-Squared with $n$ degrees of freedom is $n$, as it's the sum of $n$ squared standard normals)